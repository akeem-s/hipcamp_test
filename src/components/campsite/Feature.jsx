import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FontAwesome from 'react-fontawesome';

const propTypes = {
  presence: PropTypes.bool,
  subfeatures: PropTypes.array,
  title: PropTypes.string,
};

export default class Feature extends Component {
  constructor(props){
    super(props);
    this.renderPresenceIcons = this.renderPresenceIcons.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.renderSubfeatures = this.renderSubfeatures.bind(this);
    this.state = {
      showSubfeatures: false,
    }
  }

  handleClick(){
    this.setState({
      showSubfeatures: !this.state.showSubfeatures
    })
  };

  renderPresenceIcons(presence){
    if(presence){
      return <FontAwesome className='customFaCheck customFa' name='check'/>;
    }
    return <FontAwesome className='customFaBan customFa' name='ban'/>;
  };

  renderSubfeatures(){
    const { showSubfeatures } = this.state;
    const { subfeatures } = this.props;
    let html;
    if(subfeatures.length > 0 && showSubfeatures){
      html = subfeatures.map((f)=>{
        const { title, presence } = f;
        return (
          <div className='subfeaturesContainer' key={title}>
            <p className='subfeatureTitle'>{title}</p>
            {this.renderPresenceIcons(presence)}
          </div>
        )
      })
    }
    return html;
  };

  render(){
    const { presence, subfeatures, title } = this.props;
    const hasSubfeaturesClassName = subfeatures.length > 0 ? 'hasSubfeatures' : null;
    return (
      <div className={'feature '+hasSubfeaturesClassName} onClick={()=>{this.handleClick(subfeatures)}}>
        <div className='featureDetailsContainer'>
          <p className='featureTitle' >{title}</p>
          {this.renderPresenceIcons(presence)}
        </div>
        {this.renderSubfeatures()}
      </div>
    )
  }
};

Feature.propTypes = propTypes;
