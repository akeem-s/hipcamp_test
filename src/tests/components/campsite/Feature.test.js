import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Feature from '../../../components/campsite/Feature';
import { FEATURES } from '../../../assets/camp_features.js';

describe('Feature', () => {
  let wrapper;
  const { title, presence, subfeatures } = FEATURES[0];
  const getWrapper = () => {
    const component = shallow(<Feature key={title} title={title} presence={presence} subfeatures={subfeatures}/>);
    return component;
  };

  beforeEach(() => {
    wrapper = getWrapper();
  });

  test('component renders', () => {
    expect(wrapper.length).toEqual(1);
  });

  test('receives props', () => {
    expect(wrapper.instance().props.title).toEqual(title)
  })

  test('snapshot should contain expected DOM elements', () => {
    const renderedValue = renderer.create(wrapper).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
