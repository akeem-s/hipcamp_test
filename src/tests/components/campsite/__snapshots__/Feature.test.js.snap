// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature snapshot should contain expected DOM elements 1`] = `
<div
  className="feature null"
  onClick={[Function]}
>
  <div
    className="featureDetailsContainer"
  >
    <p
      className="featureTitle"
    >
      Toilet
    </p>
    <span
      aria-hidden={true}
      className="fa fa-check customFaCheck customFa"
    />
  </div>
</div>
`;
